File: exercism/csharp/need-for-speed/NeedForSpeed.cs
using System;

class RemoteControlCar
{
    private int speed = 0;
    private int battery = 100;
    private int distance = 0;
    private int batteryDrain;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public bool BatteryDrained()
    {
        return battery < batteryDrain;
    }

    public int DistanceDriven()
    {
        return distance;
    }

    public void Drive()
    {
        if (batteryDrain > battery) {
            return;
        }
        battery -= batteryDrain;
        distance += speed;
    }

    public static RemoteControlCar Nitro()
    {
        return new RemoteControlCar(50, 4);
    }
}

class RaceTrack
{
    private int distance = 0;

    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool CarCanFinish(RemoteControlCar car)
    {
        do {
            car.Drive();
        } while (!car.BatteryDrained());

        return car.DistanceDriven() >= distance;
    }
}

File: exercism/go/need-for-speed/need_for_speed.go
package speed

type Car struct {
	speed        int
	battery      int
	batteryDrain int
	distance     int
}

// NewCar creates a new remote controlled car with full battery and given specifications.
func NewCar(speed, batteryDrain int) Car {
	return Car{speed, 100, batteryDrain, 0}
}

type Track struct {
	distance int
}

// NewTrack created a new track
func NewTrack(distance int) Track {
	return Track{distance}
}

// Drive drives the car one time. If there is not enough battery to drive on more time,
// the car will not move.
func Drive(car Car) Car {
	if car.batteryDrain > car.battery {
		return car
	}

	car.distance += car.speed
	car.battery -= car.batteryDrain

	return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
	cycle := track.distance / car.speed

	return cycle <= (car.battery / car.batteryDrain)
}
